// grammy/out/convenience/frameworks.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultAdapter = exports.adapters = void 0;
const frameworks_shared_js_1 = require("./frameworks.shared.js");
/** Node.js native 'http' and 'https' modules */
const http = (req, res) => {
    const secretHeaderFromRequest = req.headers[frameworks_shared_js_1.SECRET_HEADER.toLowerCase()];
    return {
        update: new Promise((resolve, reject) => {
            const chunks = [];
            req.on("data", (chunk) => chunks.push(chunk))
                .once("end", () => {
                const raw = Buffer.concat(chunks).toString("utf-8");
                resolve(JSON.parse(raw));
            })
                .once("error", reject);
        }),
        header: Array.isArray(secretHeaderFromRequest)
            ? secretHeaderFromRequest[0]
            : secretHeaderFromRequest,
        end: () => res.end(),
        respond: (json) => res
            .writeHead(200, { "Content-Type": "application/json" })
            .end(json),
        unauthorized: () => res.writeHead(401).end("secret token is wrong"),
    };
};
/** worktop CloudFlare workers framework */
const worktop = (req, res) => ({
    update: Promise.resolve(req.body.json()),
    header: req.headers.get(frameworks_shared_js_1.SECRET_HEADER),
    end: () => res.end(),
    respond: (json) => res.send(200, json),
    unauthorized: () => res.send(401, "secret token is wrong"),
});
/** AWS lambda serverless functions */
const awsLambda = (event, _context, callback) => ({
    update: JSON.parse(event.body),
    header: event.headers[frameworks_shared_js_1.SECRET_HEADER],
    end: () => callback(null, { statusCode: 200 }),
    respond: (json) => callback(null, { statusCode: 200, body: json }),
    unauthorized: () => callback(null, { statusCode: 401 }),
});
/** Azure Functions */
const azure = (context, req) => ({
    update: Promise.resolve(req.body),
    header: context.res.headers[frameworks_shared_js_1.SECRET_HEADER],
    end: () => (context.res = {
        status: 200,
        body: "",
    }),
    respond: (json) => {
        context.res.set("Content-Type", "application/json");
        context.res.send(json);
    },
    unauthorized: () => {
        context.res.send(401, "secret token is wrong");
    },
});
/** Next.js Serverless Functions */
const nextJs = (req, res) => ({
    update: Promise.resolve(req.body),
    header: req.headers[frameworks_shared_js_1.SECRET_HEADER],
    end: () => res.end(),
    respond: (json) => res.status(200).json(json),
    unauthorized: () => res.status(401).send("secret token is wrong"),
});
/** Sveltekit Serverless Functions */
const sveltekit = ({ request }) => {
    var _a;
    let resolveResponse;
    return {
        update: Promise.resolve(request.json()),
        header: (_a = request.headers.get(frameworks_shared_js_1.SECRET_HEADER)) !== null && _a !== void 0 ? _a : undefined,
        end: () => {
            if (resolveResponse)
                resolveResponse(new Response());
        },
        respond: (json) => {
            if (resolveResponse) {
                const res = new Response(json, {
                    headers: { "Content-Type": "application/json" },
                });
                resolveResponse(res);
            }
        },
        unauthorized: () => {
            if (resolveResponse) {
                const res = new Response("secret token is wrong", {
                    status: 401,
                });
                resolveResponse(res);
            }
        },
        handlerReturn: new Promise((resolve) => {
            resolveResponse = resolve;
        }),
    };
};
// please open a PR if you want to add another
exports.adapters = {
    http,
    https: http,
    worktop,
    "aws-lambda": awsLambda,
    azure,
    "next-js": nextJs,
    sveltekit,
    ...frameworks_shared_js_1.adapters,
};
exports.defaultAdapter = "express";
